packages/app-store/ouraring/lib/CalendarService.ts

```diff
@@ -0,0 +1,180 @@
+import { z } from "zod";
+
+import dayjs from "@calcom/dayjs";
+import { handleErrorsJson } from "@calcom/lib/errors";
+import logger from "@calcom/lib/logger";
+import prisma from "@calcom/prisma";
+import type {
+  Calendar,
+  EventBusyDate,
+  IntegrationCalendar,
+  NewCalendarEventType,
+} from "@calcom/types/Calendar";
+import type { CredentialPayload } from "@calcom/types/Credential";
+
+import getParsedAppKeysFromSlug from "../../_utils/getParsedAppKeysFromSlug";
+
+export type OuraRingAuthCredentials = {
+  access_token: string;
+  refresh_token: string;
+  expires_in: number;
+};
+
+type OuraRingReturnedData = {
+  day: string;
+  score: number;
+};
+
+const toUrlEncoded = (payload: Record<string, string>) =>
+  Object.keys(payload)
+    .map((key) => key + "=" + encodeURIComponent(payload[key]))
+    .join("&");
+
+const ouraRingAppKeysSchema = z.object({
+  client_id: z.string(),
+  client_secret: z.string(),
+});
+
+export const getOuraRingAppKeys = async () => {
+  return getParsedAppKeysFromSlug("ouraring", ouraRingAppKeysSchema);
+};
+
+const refreshTokenResponseSchema = z.object({
+  access_token: z.string(),
+  expires_in: z
+    .number()
+    .transform((currentTimeOffsetInSeconds) => Math.round(+new Date() / 1000 + currentTimeOffsetInSeconds)),
+  refresh_token: z.string().optional(),
+});
+
+export default class OuraRingCalendarService implements Calendar {
```

Comment:
Usual OAuth2 flow to check existing `access_token` or generates a new one relying on `refresh_token`, nothing much to see here.